File: js/money_test.js
const assert = require("assert")

class Money {
    constructor(amount, currency) {
        this.amount = amount
        this.currency = currency
    }

    times(multiplier) {
        return new Money(this.amount * multiplier, this.currency)
    }

    divide(divisor) {
        return new Money(this.amount / divisor, this.currency)
    }
}

class Portfolio {
    constructor() {
        this.moneys = []
    }

    add(...moneys) {
        this.moneys = this.moneys.concat(moneys)
    }

    evaluate() {
        let total = this.moneys.reduce((sum, money) => {
            return sum + money.amount
        }, 0)
        return new Money(total, "USD")
    }
}


let tenDollars = new Money(10, 'USD')
let fiveDollars = new Money(5, 'USD')
let fifteenDollars = new Money(15, 'USD')
let portfolio =  new Portfolio()
portfolio.add(fiveDollars, tenDollars)
assert.deepStrictEqual(portfolio.evaluate('USD'), fifteenDollars)


// = TEST Case
// let fiver = new Money(5, "USD")
// let tenner = fiver.times(2)
// assert.strictEqual(tenner.amount, 10)
// assert.strictEqual(tenner.currency, "USD")
// assert.deepStrictEqual(tenner, new Money(10, "USD"))
// assert.deepStrictEqual(fiver.times(2), tenner)


// = Test Case
// let originalMoney = new Money(4002, "KRW")
// let actualMoneyAfterDivision = originalMoney.divide(4)
// let expectedMoneyAfterDivision = new Money(1000.5, "KRW")
// assert.deepStrictEqual(actualMoneyAfterDivision, expectedMoneyAfterDivision)

File: py/money_test.py
import unittest
import functools
import operator


class Dollar:
    def __init__(self, amount):
        self.amount = amount

    def times(self, multiplier):
        return Dollar(self.amount * multiplier)


class Money:
    def __init__(self, amount, currency):
        self.amount = amount
        self.currency = currency

    def times(self, multiplier):
        return Money(self.amount * multiplier, self.currency)


class Portfolio:
    def __init__(self):
        self.moneys = []

    def add(self, *moneys):
         self.moneys.extend(moneys)

    def evaluate(self, currency):
         total = functools.reduce(operator.add, map(lambda m: m.amount, self.moneys), 0)
         return Money(total, currency)


class TestMoney(unittest.TestCase):
    def testMultiplication(self):
        fiver = Dollar(5)
        tenner = fiver.times(2)
        self.assertEqual(10, tenner.amount)

    def testMultiplicationInEuros(self):
        tenEuros = Money(10, "EUR")
        twentyEuros = tenEuros.times(2)
        self.assertEqual(20, twentyEuros.amount)
        self.assertEqual("EUR", twentyEuros.currency)

    def testAddition(self):
        fiveDollars = Money(5, 'USD')
        tenDollars = Money(10, 'USD')
        fifteenDollars = Money(15, 'USD')
        portfolio = Portfolio()
        portfolio.add(fiveDollars, tenDollars)
        print(fifteenDollars, portfolio.evaluate('USD'))
        self.assertEqual(fifteenDollars, portfolio.evaluate('USD'))

if __name__ == '__main__':
    unittest.main()